mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantCast.cpp

```diff
@@ -0,0 +1,171 @@
+    // outside of the target integer range.
+    auto floatSize = APFloat::getSizeInBits(toConvert.getSemantics());
+    APSInt converted(std::max(floatSize, targetWidth), false);
+    bool ignored = false;
+    toConvert.convertToInteger(converted, APFloat::rmNearestTiesToEven,
+                               &ignored);
+    // Clip to allowed range.
+    if (targetWidth < floatSize) {
+      return converted.truncSSat(targetWidth);
+    }
+    return converted;
+  }
+
+  static APInt convertIntToInt(const APInt &toConvert, IntegerType targetType) {
+    // Make sure to properly translate booleans
+    if (targetType.getWidth() == 1) {
+      return toConvert.isZero() ? APInt::getZero(1) : APInt::getAllOnes(1);
+    }
+    return toConvert.sextOrTrunc(targetType.getIntOrFloatBitWidth());
+  }
+
+  static void warnAboutNaNToIntCast(DenseElementsAttr elements, CastOp location,
+                                    PatternRewriter &rewriter) {
+    // This is only relevant if the input values are float
+    if (!isa<FloatType>(elements.getElementType())) {
+      return;
+    }
+    // Check if it is an float to integer conversion
+    auto resultType = location.getOutput().getType();
+    if (!isa<IntegerType>(cast<TensorType>(resultType).getElementType())) {
+      return;
+    }
+
+    for (auto val : elements.getValues<APFloat>()) {
+      // Report encountered NaNs once
+      if (val.isNaN()) {
+        location->emitWarning(
+            "Float tensor is casted to integer and it contains NaN values. The "
+            "cast results in an unspecified value.");
+        return;
+      }
+    }
+  }
+
+  LogicalResult matchAndRewrite(CastOp tosaCast,
+                                PatternRewriter &rewriter) const override {
+    auto inputTensor = tosaCast.getInput();
+
+    // If the input tensor is not constant, we cannot fold it.
+    auto isDenseConst =
+        notifyIfNoTosaDenseConstantTensor(inputTensor, tosaCast, rewriter);
+    if (failed(isDenseConst)) {
+      return isDenseConst;
+    }
+
+    auto fromType = inputTensor.getType().getElementType();
+    auto toType = tosaCast.getOutput().getType().getElementType();
+
+    DenseElementsAttr elements;
+    matchPattern(inputTensor, m_Constant(&elements));
+
+    // Issue a warning if we convert float -> int and NaNs are present; the
+    // result value is unspecified in that case
+    warnAboutNaNToIntCast(elements, tosaCast, rewriter);
+
+    // Only fold splat tensors and those used only once to avoid duplicating
+    // them.
+    if (!inputTensor.hasOneUse() && !isa<SplatElementsAttr>(elements)) {
+      return rewriter.notifyMatchFailure(tosaCast,
+                                         "Currently, casts will only be folded "
+                                         "if its input only has a single user");
+    }
+
+    DenseElementsAttr res;
+    if (auto intOutTy = dyn_cast<IntegerType>(toType)) {
```

Comment:
* Should this rather be nested switches?
* Is it ok to use assertions here? The spec doesn't allow for other types, so non of them should be able to fail unless the spec changes.

Reply:
Addressed in 4659d22ac9393c57d902d00c4b7091b9384ee293.